aten/src/ATen/core/ivalue_inl.h

```diff
@@ -468,11 +468,21 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
     auto fut = createInstance(std::move(type));
     addCallback(
         [fut, cb = std::move(callback)]() {
+          // Get the true stacktrace if call to markCompleted() fails. We cannot
+          // simply call fut->markCompleted(cb()) since if markCompleted fails
+          // after setting completed_ flag, setError will report error that
+          // completed_ is set, instead of the true error.
+          at::IValue cbValue;
+          bool hasException = false;
           try {
-            fut->markCompleted(cb());
+            cbValue= cb();
           } catch (std::exception&) {
+            hasException = true;
             fut->setError(std::current_exception());
           }
+          if (!hasException) {
+            fut->markCompleted(std::move(cbValue));
```

Comment:
Yeah the reason I suggested catching exceptions here and fataling is that it's not fully clear to me what would happen to uncaught exceptions. I suspect they would "bubble up" to whoever called `markCompleted`, which might be a random thread in the thread pool, which may even be configured to catch, log and then ignore exceptions. I'm afraid that people don't watch logs unless something really goes wrong but, in this case, the "only" thing that would happen is an incorrect CUDA synchronization, which is a silent correctness issue. I'd rather we notify the user about such an issue as soon as we can, even if it means being very harsh.

Reply:
Yeah, I think it would just show up in the logs. Regarding the thread pool, I believe that the c10 thread pool swallows and logs: https://github.com/pytorch/pytorch/blob/master/c10/core/thread_pool.cpp#L112, so it would only be surfacing to the user via logging as you mentioned.